Fix displaying wrong membership on "My Team" view.

peter-fogg opened the pull request:
## [TNL-3410](https://openedx.atlassian.net/browse/TNL-3410)

The heart of this fix is to change the "My Team" view to use a
collection of teams instead of memberships. The team card is
refactored to only take a team, rather than attempt to be polymorphic
over teams and team memberships. This change enabled removing a good
amount of old code. This also requires adding a `username` parameter
to the teams list endpoint which allows getting a list of all teams
for a single user, in order to allow refreshing the "My Teams" view
correctly. Currently this list is of length at most one.

Unfortunately I made this change around the window of the DRF upgrade, so I need a hairy rebase onto master... There's also one failing JS test that I'm working on. I haven't yet added the `username` parameter to the API documentation wiki page, but I'll get to that once review is done.

- [x] Document `username` parameter in Teams API wiki page

### Sandbox
- [x] http://peter-fogg.sandbox.edx.org/courses/course-v1:TestX+T101+2015/teams/#my-teams

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @andy-armstrong
- [x] Code review: @cahrens

peter-fogg commented:
@cahrens @andy-armstrong I've rebased onto master (and learned about serializers in DRF 3!).

cahrens commented:
@peter-fogg there are some test failures

peter-fogg commented:
@cahrens Test failures should be fixed now.

cahrens commented:
@peter-fogg still a few

peter-fogg commented:
@cahrens They're fixed now.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/topic_teams_spec.js:
```diff
@@ -1,21 +1,21 @@
                 model: TeamSpecHelpers.createMockTopic(),
                 collection: options.teams || TeamSpecHelpers.createMockTeams(),
-                teamMemberships: options.teamMemberships || TeamSpecHelpers.createMockTeamMemberships(),
+                canUserCreateTeam: function () {
```
It's sad to have to duplicate this method in the tests. Is there no reasonable way to create a utility method?

andy-armstrong replied:
I agree. I think that the method should live in ``TopicTeamsView`` itself, and that it should just be changed to receive the ``myTeamsCollection``.

peter-fogg replied:
As noted below, fixed.

cahrens commented on lms/djangoapps/teams/static/teams/js/spec_helpers/team_spec_helpers.js:
```diff
@@ -290,7 +265,8 @@ define([
         teamsDetailUrl: '/api/team/v0/teams/team_id',
         teamMembershipsUrl: '/api/team/v0/team_memberships/',
-        teamMembershipDetailUrl: '/api/team/v0/team_membership/team_id,' + testUser,
+        teamMembershipDetailUrl: '/api/team/v0/teams/',
```
I am confused about the names here. I don't really like renaming teamMembershipDetailUrl as leaveTeamUrl, as it isn't just for leaving teams (though that may be all we are using). It's for interacting with a team membership, with the DELETE method doing deletion. I would prefer to keep the name as it was.

And then I'm confused why /api/team/v0/teams/ should be teamMembershipDetailUrl, as you are using it to get all team memberships for a given user, correct? Perhaps myTeamsUrl?

peter-fogg replied:
That makes a lot more sense; I'll make that change.

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -153,14 +153,15 @@ def get(self, request, course_id):
+
+        user_teams = CourseTeam.objects.filter(membership__user=user)
+        user_teams_data = self._serialize_and_paginate(
             MembershipPagination,
```
Should it still be using MembershipPagination?

peter-fogg replied:
Yes, since `MembershipPagination` controls the page size. That name is confusing, though; I'll change it.

cahrens commented:
@peter-fogg I don't see any test coverage for the change (a test that fails without your fix). I think a bok choy test would be best for this.

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -414,6 +418,10 @@ def get(self, request):
 
+        username = request.QUERY_PARAMS.get('username', None)
+        if username is not None:
+            result_filter.update({'membership__user__username': username})
```
There is no unit test coverage for this change.

peter-fogg replied:
Added a test in `test_views`.

peter-fogg commented:
@cahrens Ah yes, I did get caught up in all the other tests and forgot to write a new one. I will do that.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/spec/views/my_teams_spec.js:
```diff
@@ -1,39 +1,30 @@
-                },
-                options
-            )).render();
+        var createMyTeamsView = function(teams) {
```
Nit: rename the parameter to ``myTeams``

peter-fogg replied:
Done.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/spec/views/my_teams_spec.js:
```diff
@@ -41,36 +32,37 @@ define([
                     page : '1',
-                    page_size : '10',
-                    text_search: ''
+                    page_size : '2',
```
Why did you decrease the page size? It shouldn't matter, right? Unless you are testing something about page sizes, it would seem clearer to leave it the same default as the feature uses.

peter-fogg replied:
`TEAM_MEMBERSHIPS_PER_PAGE` is defined as 2 in `teams/views.py` -- I'm not sure why the original test was testing with 10 per page, but this is the correct value.

andy-armstrong replied:
Thanks for the explanation. 2 seems a strange value, but your change makes sense given that.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -351,12 +347,21 @@
                             model: topic,
                             collection: collection,
-                            teamMemberships: this.teamMemberships,
+                            canUserCreateTeam: function () {
```
I find it strange to have this definition defined here. I think the reason why this isn't defined in TopicTeamsView is that it doesn't have access to the myTeamsCollection, so maybe it should just be passed to it. Is there a use case for two different TopicTeamsViews having different implements of ``canUserCreateTeam``?

peter-fogg replied:
I don't think so; I've changed this.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -16,38 +16,35 @@
-                                self.$el.append(_.template(teamActionsTemplate, {message: message}));
-                            }
-                        });
+                    this.collection.refresh().done(function() {
```
Much cleaner!

andy-armstrong commented on lms/djangoapps/teams/templates/teams/teams.html:
```diff
@@ -42,6 +42,7 @@
         teamsDetailUrl: '${ teams_detail_url }',
         teamMembershipsUrl: '${ team_memberships_url }',
         teamMembershipDetailUrl: '${ team_membership_detail_url }',
+        leaveTeamUrl: '${leave_team_url}',
```
Nit: all the other usages here have spaces around the brackets, so you should probably stay consistent (I personally prefer not to include the whitespace)

peter-fogg replied:
I changed this to include the whitespace.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/topic_teams.js:
```diff
@@ -68,7 +65,9 @@
                     event.preventDefault();
-                    Backbone.history.navigate('topics/' + this.model.id + '/create-team', {trigger: true});
+                    Backbone.history.navigate(
+                        'topics/' + this.model.id + '/create-team', {trigger: true}
```
Nit: I'd put the second parameter on a new line as it is a little hidden on the same line.

peter-fogg replied:
Done.

andy-armstrong commented:
Thanks for tackling this clean up, @peter-fogg. It is so much easier to understand now. I'm done with my review so let me know when you've addressed our feedback.

peter-fogg commented:
@cahrens @andy-armstrong I believe I've addressed your comments. I've added a Bok Choy test asserting that the number of team members is displayed correctly.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/teams_tab.js:
```diff
@@ -95,26 +95,22 @@
-                            privileged: this.context.userInfo.privileged,
-                            staff: this.context.userInfo.staff,
-                            parse: true
+                            per_page: 10,
```
I'm confused again. Didn't you say this was 2? Or is the server default 2, but the client is passing 10?

peter-fogg replied:
You're right, I missed setting this correctly. Fixed now.

andy-armstrong commented:
:+1: looks great!

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -339,6 +348,12 @@ def test_member_of_a_team(self):
 
+    def test_multiple_team_members(self):
+        teams = self.create_teams(self.topic, 1)
+        self.create_memberships(4, teams[0]['id'])
```
How does the active user get enrolled in the course? Through create_teams (in which case there should be 5 members, not 4)? Or somehow magically when creating 4 memberships via create_memberships?

peter-fogg replied:
The users are all enrolled through `create_memberships`.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -339,6 +348,12 @@ def test_member_of_a_team(self):
             self.my_teams_page.visit()
         self.verify_teams(self.my_teams_page, teams)
 
+    def test_multiple_team_members(self):
```
If you have time, can you add a flaky decorator to this and run the new test 10 times? This is something the team wants to do with all new tests to help avoid introducing new flaky tests.

peter-fogg replied:
I've done this, and it passed 10 times.

cahrens commented on common/test/acceptance/tests/lms/test_teams.py:
```diff
@@ -339,6 +348,12 @@ def test_member_of_a_team(self):
         self.verify_teams(self.my_teams_page, teams)
 
+    def test_multiple_team_members(self):
+        teams = self.create_teams(self.topic, 1)
```
Please add a BDD comment describing the test case (following the pattern of other test cases in the file).

cahrens commented on lms/djangoapps/teams/static/teams/js/spec/views/my_teams_spec.js:
```diff
@@ -41,36 +32,37 @@ define([
+                teams = TeamSpecHelpers.createMockTeams({
+                    results: []
+                }, {
+                    per_page: 2,
```
Is there any reason why per_page is 2 here?

cahrens commented on lms/djangoapps/teams/views.py:
```diff
@@ -107,8 +107,8 @@ class TopicsPagination(TeamAPIPagination):
 
-class MembershipPagination(TeamAPIPagination):
-    """Paginate memberships. """
+class MyTeamsPagination(TeamAPIPagination):
```
Thanks for the rename-- much clearer!

cahrens commented:
:+1:
